### ISSUE TITLE
[Question] The case for Orleans streams

### ISSUE THREAD

**Issue by RamanBut-Husaim** [dotnet/orleans#5164]
Hello colleagues,

I'm evaluating the usage of `orleans` at the project and would like to clarify a couple of things about `orleans` streams design. I'm familiar with actor model in general (some BEAM VM background with Erlang/Elixir) but still completely new with `orleans` as only went through the doc. Sadly there is no graphical schemas in the docs that could make them more clear.

## Intro 
What I'm trying to do it to understand `orleans` streams based on the practical task. We have a bunch of microservices in the system that will be responsible for producing events of different types. Number of event types are limited (~ 20). The goal is to build a notification system (webhooks) where clients could subscribe to certain types of events.

## Parties

Producers - generate events of certain types, the number is limited is will be < 20;
Events - limited number of event types (~20);
Grains - represents subscriptions (consumers) where each subscription could listen to one or more event types and all of them are of the same grain type; the number of subscriptions is much bigger (hundreds and thousands);

### Case 1 Topology

![event_stream_1](https://user-images.githubusercontent.com/5213282/48087607-d7a3f200-e210-11e8-8036-594bded9e978.png)

All grains are of the same type and stateful. In this case each grain (1, 2, ..., n) will subscribe to certain event types. As far as I understand from the documentation the subscription handles that are used for resuming subscriptions are defined on the stream level. As a result in order to continue the processing correctly as part of the activation the grain should save subscription handles somewhere in its state and use them instead of calling `GetAllSubscriptionHandles`.

**Question**: Is this a correct understanding or subscription handles depend on the grain instance as well - somehow capture the grain instance identity (1, 2, .., n) implicitly and `GetAllSubscriptionHandles` will return handles for this particular stream + grain instance?

Besides I have an impression that such approach where we have a lot of consumers and limited number of streams does not play well with the design of `orleans` streams. As a result I've come to the case 2.

### Case 2 Topology

![event_stream_2](https://user-images.githubusercontent.com/5213282/48088436-0cb14400-e213-11e8-866a-7f5797adfa04.png)

There are two types of grains:
+ yellow - subscription processing grains that are responsible for event processing; each such grain has a particular stream that contain events of different types in which this grain is interested in;
+ grey - distributors - they gather events of a particular type and distribute them across yellow grain streams;

With such approach we have a limited number of grey distributor grains that push events to a large number of streams for processing. This seems to be the case for `orleans` streams. There will be no issue with `GetAllSubscriptionHandles` as each stream will be unique for a subscription processing grain instance (1, 2, ..., n). 

What bothers me here:
+ each grey grain will have to keep a mapping between event type and yellow grain stream identity and this could become potentially a big thing; as far as I understand this could be distributed more with some sort of sharding;
+ events will have more hops as it will first trap into a distributor and next from distributor to the subscription processing stream;

**Question**: Is my understand correct and `orleans` is primarily designed to have a lot of relatively small streams and in cases where we have a small number of event types and large number of consumers we need to distribute such work by creating a lot of streams?

Thanks.

**Comment by sergeybykov** [comment]
>Question: Is this a correct understanding or subscription handles depend on the grain instance as well - somehow capture the grain instance identity (1, 2, .., n) implicitly and GetAllSubscriptionHandles will return handles for this particular stream + grain instance?

Yes, `GetAllSubscriptionHandles` will return all subscriptions of the calling grain to the given stream. So in most cases it'll likely return a single handle per stream the grain is subscribed to.

>Besides I have an impression that such approach where we have a lot of consumers and limited number of streams does not play well with the design of orleans streams. As a result I've come to the case 2.

In principle, there's nothing wrong in this approach. However, a high cardinality of fan-out for each event might lead to higher messaging load and lower throughput (very chatty communication). Depends on the load pattern of course.

>Question: Is my understand correct and orleans is primarily designed to have a lot of relatively small streams and in cases where we have a small number of event types and large number of consumers we need to distribute such work by creating a lot of streams?

I think the answer depends on the scale factors here. Orleans streaming is designed to be most efficient when there are enough physical queues (that vistual streams are mapped to) to distributed the work of pulling events from them (pulling agents do that) across silos in the cluster. By default, we create N/2 pulling agents per stream provider per silo, where N is the number of CPU cores. Because of that, if the cardinalities of streams and queues are too small, that might lead to uneven distribution of pulling work across the silos.

Your Case 2 topology reminds me of the streaming architecture - the gray grains can be viewed as pulling agents that receive events from the "queues" (even though they are actually streams) and fan them out to subscribed yellow grains, except instead of doing that with direct RPC calls they deliver events to them via another set of queues.

So it's not clear to me what you are gaining here other than maybe simplifying recovery from failures with each yellow grain recovering in total isolation from other yellow grains. The cost of that is the extra roundtrips to queues. This could be a reasonable tradeoff, just hard to tell upfront more details about the actual workload.

Have you considered leveraging implicit subscriptions? It seems to me you could do that for both gray and yellow grains. With implicit subscriptions recovery is even simpler, as the pubsub is not used and subscription handles won't need to be saved or retrieved.

**Comment by RamanBut-Husaim** [comment]
@sergeybykov Hello. Huge thanks for the provided information. 

> Yes, GetAllSubscriptionHandles will return all subscriptions of the calling grain to the given stream. So in most cases it'll likely return a single handle per stream the grain is subscribed to.

Just to clarify. Let's take a look at case 1, grain 1 and grain 2 which are connected to stream k (green) and stream 2 (red).
**Streams**
+ stream k (green) - identified by guid 1 and namespace 1;
+ stream 2 (red) - identified by guid 2 and namespace 2;
**Grains**
+ Grain 1 is subscribed to stream k (green) and stream 2 (red), for stream k - handle with id 23, for stream 2 with id 48;
+ Grain 2 is subscribed to stream k (green) only and upon subscription it will receive a handle with id 84 (different from handle 23 for Grain 1);

If I understand correctly when Grain 1 calls `GetAllSubscriptionHandles` he will receive handles with ids 23 and 48 only (not id=84) even though both Grain 1 and Grain 2 are subscribed to the same stream k (green). For Grain 2 only handle with id 84 will be returned. 

Is this correct? 

---

> Orleans streaming is designed to be most efficient when there are enough physical queues (that vistual streams are mapped to) to distributed the work of pulling events from them (pulling agents do that) across silos in the cluster.

I'm curious how physical queues are actually map to the streams and stream providers. Is there one physical queue per stream provider or the correlation is different?

And do I understand correctly that increasing the number of stream providers on the silo could be a bad idea as pulling agents per stream provider will fight for CPU resources? 

---

> So it's not clear to me what you are gaining here other than maybe simplifying recovery from failures with each yellow grain recovering in total isolation from other yellow grains. 

Yep, the idea behind this is to make them as much isolated as possible + make the understanding as clear as possible. Each yellow grain (webhook) will receive events from a single stream only. 

---
> Have you considered leveraging implicit subscriptions?

Actually I'm going to use implicit subscription to save all events into the database :). Regarding grey and yellow. As far as I can see it won't be possible due to the following:
+ Gray Grains. Though we will have a limited number of event types they are dynamic - it means that the system does not know their names up front and receives them from front-end (wep api); implicit subscription relies on compile-time attribute names; I've noticed that there are `subscriptionnamespaceprovider` interface that could potentially help to solve the problem though it is unclear how we could get streams to subscribe to inside the grain itself using pattern :); `GetStream` seems to accept only guid and exact namespace;
+ Yellow Grains. The events to which each such grain is subscribed is dynamic as well (webhook) and could be changed via front-end (wep api). As a result explicit subscription seem to be the choice here.

**Comment by sergeybykov** [comment]
>If I understand correctly when Grain 1 calls GetAllSubscriptionHandles he will receive handles with ids 23 and 48 only (not id=84) even though both Grain 1 and Grain 2 are subscribed to the same stream k (green). For Grain 2 only handle with id 84 will be returned.

Note that `GetAllSubscriptionHandles()` is available on a stream. So Grain 1 will need to call it twice, once for the green stream and once for the red one. Grain 2 will indeed receive back a handle for the green stream that is different from the handle that Grain 1 will get.

>I'm curious how physical queues are actually map to the streams and stream providers. Is there one physical queue per stream provider or the correlation is different?

By default it's a hash-based mapping - `HashRingBasedStreamQueueMapper`. A custom one could be configured instead. Queues are preconfigured (e.g. EventHub partitions), and the mapper is responsible for mapping streams to them.

>And do I understand correctly that increasing the number of stream providers on the silo could be a bad idea as pulling agents per stream provider will fight for CPU resources?

The number of pulling agents per silo is configurable. So you can fine-tune it. But I don't expect a big need for that because all pulling agents execute on the same threads as grains and don't started their own threads.

**Comment by RamanBut-Husaim** [comment]
@sergeybykov Hello. Thanks a lot for this information.

Sergey, are there any whitepapers or detailed doc available on the `orleans` streaming part? Or only source code could server as a guide for this?

I know about official [orleans whitepaper](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/Orleans-MSR-TR-2014-41.pdf) but seems that it does not contain a lot about streams.

**Comment by sergeybykov** [comment]
I'm afraid the [streaming docs](http://dotnet.github.io/orleans/Documentation/streaming/index.html) and source code is all we have at the moment.
